gui/default/syncthing/folder/ignoresService.js

```diff
@@ -0,0 +1,257 @@
+                var data = response.data;
+                return (data.ignore || []).map(parsePattern);
+            });
+        }
+
+        function parsePattern(line) {
+            var stripResult = splitPrefix(line.trim());
+            var prefixes = stripResult[0];
+            var hasPrefix = prefixes['(?i)'] || prefixes['(?d)'];
+            var path = toPath(stripResult[1]);
+            var matchFunc = !hasPrefix && chooseMatcher(path);
+
+            return {
+                text: line,
+                isSimple: !!matchFunc,
+                isNegated: prefixes['!'],
+                path: path,
+                matchFunc: matchFunc,
+            };
+        };
+
+        // Adapted from lib/ignore/ignore.go@parseLine
+        function splitPrefix(line) {
+            var seenPrefix = { '!': false, '(?i)': false, '(?d)': false };
+
+            var found = true;
+            while (found === true) {
+                found = false;
+                for (var prefix in seenPrefix) {
+                    if (!seenPrefix[prefix] && line.indexOf(prefix) === 0) {
+                        seenPrefix[prefix] = true;
+                        line = line.slice(prefix.length);
+                        found = true;
+                        break;
+                    }
+                }
+            }
+
+            return [seenPrefix, line];
+        }
+
+        // Infer a path from wildcards that might be used to match a file path
+        // by prefix (for simple patterns)
+        // The changes made to the path make it more specific than the original
+        // pattern. It is useful for prefix matcher. Should a glob matcher be
+        // added this path should not be used to match because it does not
+        // represent the complete pattern. The most common effect of the changes
+        // is matching patterns like `*`, `**`, or `/*` without a special case
+        function toPath(line) {
+            // Add a leading slash when the pattern begins with wildcards
+            // because a pattern beginning with * or ** will match entries at
+            // the root (as well as those in child directories)
+            line = line.replace(/^\*+/, '/$&'); // `$&` inserts the matched substring
+
+            // Trim trailing wildcards after separator because child paths would
+            // already be prefixed by this path
+            line = line.replace(/\/\*+$/, '/');
+            return line;
+        }
+
+        // Inspect the pattern to determine if it is simple enough to match,
+        // then return a function that can be applied to match a path.
+        function chooseMatcher(line) {
+            if (line.length === 0) return neverMatch;
+            if (line.indexOf('//') === 0) return neverMatch; // comment
+            if (line.indexOf('/') !== 0) return null; // not a root line
+            if (line.length > 1 && line.charAt(line.length - 1) === '/') return null; // trailing slash
```

Comment:
There is also `#include` directive that doesn't seem to be handled here.

Reply:
Also, we have crazy stuff like:
```
foo/*/bar                   # any directory at the same level
```
or even
```
pics/**/Thumbs.db           # any directory at any level
```
Which probably needs to abort if it sees it?

Reply:
The two wildcard examples are handled on lines 205-211. After removing special characters that have been escaped `\*` we check for special characters including `*`. If any are present, returning null indicates this is an advanced pattern and we don't know how to match it to files.

There is an important distinction between the neverMatch function and null (no function): the former is for a pattern our UI can safely work around as we add and remove patterns, the latter is for a pattern too complex to render or modify with the UI. Presence of advanced patterns is used in syncthingController.js to determine whether to disable the UI entirely, because adding patterns on the wrong side of a glob could have unintended effects.

This is an ugly function. For now, I'm not sure of a simpler way to inspect the pattern than a series of `if`s for each check. I hope the comments and tests (ignoreService.test.js:264 in this case) can help.

Reply:
Regarding #include, I think we have to disable the browse UI when it is present.

I was curious whether the API offers enough information to render it in some way. The rest/db/ignores endpoint reflects the patterns from the file in the `expanded` field, but `ignore` field contains only the text representation "#include patterns.txt". Perhaps if there was a way to link the #include line to the patterns inside, we could show the effects and... disable modifications to paths affected by the #included file? Insert new patterns before the #include command? Not sure. This is speculation about something to support in the future - for now an "#include" command will be parsed as an advanced pattern and have the same effects as a glob pattern.